test/opt/assembly_builder.h

```diff
@@ -0,0 +1,207 @@
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef LIBSPIRV_TEST_OPT_ASSEMBLY_BUILDER
+#define LIBSPIRV_TEST_OPT_ASSEMBLY_BUILDER
+
+#include <algorithm>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace spvtools {
+
+// A simlple SPIR-V assmebly code builder for test uses. It builds an SPIR-V
+// assembly module from vectors of assembly strings. It allows users to add
+// instructions to the main function and the type-constants-globals section
+// directly. It relies on OpName instructions and friendly-name disassembling
+// to keep the ID names unchanged after assembling.
+//
+// An assembly module is divided into several sections:
+//  Global Preamble:
+//    declare capabilities, extensions, entry points, etc.
+//  Names:
+//    declare the names for IDs with OpName instructions.
+//  Types, Constants and Global variables:
+//    declare types, constants and global variables.
+//  Main Function:
+//    contains instructions in main function, just after OpLable for the entry
+//    block.
+//  Main Function Postamble:
+//    contains the return and fucntion end instruction.
+//
+// The assembly code is built by concatenating all the strings in the above
+// section.
+//
+// Users define the contents in section <Type, Constants and Global Variables>
+// and <Main Function>. The <Names> section is to hold the names for IDs to
+// keep them unchanged before and after assembling. All defined IDs to be added
+// to this code builder will be assigned with a global name through OpName
+// instruction. The name is extracted from the definition instruction.
+//  E.g. adding instruction: %var_a = OpConstant %int 2, will also add an
+//  instruction: OpName %var_a, "var_a".
+//
+// Note that the name must not be used on more than defined IDs and friendly-name
+// disassembling must be enabled so that OpName instructions will be respected.
+class AssemblyBuilder {
+ public:
+  // Initalize an minimal SPIR-V assembly code as the template. The minimal
+  // contains an empty main function and some predefined names for the main
+  // function.
+  AssemblyBuilder()
+      : global_preamble_vec_({
+            // clang-format off
+                  "OpCapability Shader",
```

Comment:
This seems rather specific.  But it's ok because it's all you've needed.  If we need to, we can always refactor later.